node/src/components/storage.rs

```diff
@@ -1352,42 +1352,12 @@ impl Storage {
         Ok(blocks)
     }
 
-    /// Returns the vector of deploys whose TTL hasn't expired yet.
-    fn get_finalized_deploys(
-        &self,
-        ttl: TimeDiff,
-    ) -> Result<Vec<(DeployHash, DeployHeader)>, FatalStorageError> {
+    /// Returns the vector of blocks that could still have deploys whose TTL hasn't expired yet.
+    fn get_finalized_blocks(&self, ttl: TimeDiff) -> Result<Vec<Block>, FatalStorageError> {
         let mut txn = self.env.begin_ro_txn()?;
         // We're interested in deploys whose TTL hasn't expired yet.
-        let ttl_expired = |block: &Block| block.timestamp().elapsed() < ttl;
-        let mut deploys = Vec::new();
-        for block in self.get_blocks_while(&mut txn, ttl_expired)? {
-            for deploy_hash in block
-                .body()
-                .deploy_hashes()
-                .iter()
-                .chain(block.body().transfer_hashes())
-            {
-                let deploy_header = match self.get_deploy_header(&mut txn, deploy_hash)? {
-                    Some(deploy_header) => deploy_header,
-                    None => {
-                        let deploy_hash = deploy_hash.to_owned();
-                        return Err(FatalStorageError::NonExistentDeploy {
-                            block: Box::new(block),
-                            deploy_hash,
-                        });
-                    }
-                };
-                // If block's deploy has already expired, ignore it.
-                // It may happen that deploy was not expired at the time of proposing a block but it
-                // is now.
-                if deploy_header.timestamp().elapsed() > ttl {
-                    continue;
-                }
-                deploys.push((*deploy_hash, deploy_header));
-            }
-        }
-        Ok(deploys)
+        let ttl_not_expired = |block: &Block| block.timestamp().elapsed() < ttl;
```

Comment:
Not all deploys returned will have their TTL non-expired yet. Is that intentional and OK?

Reply:
Yes, that's fine:
`finalized_deploys` is a filter to prevent replays. It wouldn't hurt if it contained _all_ finalized deploys, but to preserve space we can remove the ones where we are sure that they are expired.

Reply:
> but to preserve space we can remove the ones where we are sure that they are expired.

Yes, that's what I was thinking about. Not a blocker but a nice-to-have.

Reply:
But we are already doing that! When pruning, we remove the expired entries.

Reply:
Seems incidental. We are interested in deploy hashes (as PR description says) but we are pulling in whole blocks instead. Maybe there should be `get_finalized_deploy_hashes` endpoint instead.

Reply:
To me the current approach seems more natural: Returning whole blocks instead of just a list of deploy hashes together with an expiry date estimate. But if you prefer it I'd be fine with the alternative, too.

Reply:
No, it's your PR and it does the job so I'm fine. I was just trying to understand why it wasn't clear to me what's going on since the beginning.